### PR TITLE
Refactor `opt-dist` to simplify local building

### PR DESCRIPTION
**Pull request by Kobzol**
This PR refactors the `opt-dist` tool to make it easier to invoke it locally, outside of CI, and thus simplify building PGO/BOLT optimized `rustc` builds e.g. for distro maintainers. It should also make it easier to run the PGO/BOLT workflow locally e.g. to profile performance or debug issues (looking at you, https://github.com/rust-lang/rust/pull/115554).

### CODE REVIEW

File: src/tools/opt-dist/src/environment.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
+use camino::Utf8PathBuf;
+use derive_builder::Builder;
+
+#[derive(Builder)]
```
Comment (fbstj) [inline src/tools/opt-dist/src/environment.rs:4]: if you're deriving builders, is it worth deriving getters also?

Comment (Kobzol) [reply src/tools/opt-dist/src/environment.rs:4]: I don't think that `derive_builder` allows that. I think that the current code is fine, there's just a few functions, we could even make the fields public I suppose.

**Comment by Kobzol** [comment]
@rust-timer build cee65ced60e6b28438e74bd0df75798449e85c85

**Comment by matthiaskrgr** [comment]
@bors r- rollup=never
https://github.com/rust-lang/rust/pull/115928#issuecomment-1723182203

**Comment by Kobzol** [comment]
Oops, there was an error on Windows. I'll do a Windows CI try build to make sure that it works there.

@bors try

**Comment by Kobzol** [comment]
Ok, it seems to work now.

@bors r=Mark-Simulacrum

**Pull request merged**